exporters/trace/gcp_exporter/internal/gcp_exporter.cc

```diff
@@ -0,0 +1,208 @@
+    return trace::TraceId(trace_id_buf);
+}
+
+
+/**
+ * Generates a random span id
+ * NOTE: This is subject to change in the future and will be moved to 
+ *       sdk/src/trace/span.cc
+ */
+trace::SpanId GenerateRandomSpanId()
+{
+    uint8_t span_id_buf[trace::SpanId::kSize];
+    opentelemetry::sdk::common::Random::GenerateRandomBuffer(span_id_buf);
+    return trace::SpanId(span_id_buf);
+}
+
+
+/**
+ * Sets google protobuf object's unix start time, both seconds and nanos
+ * 
+ * @param from_span - The span whose start time we extract and convert to google protobuf timestamp
+ * @param proto - A google protobuf timestamp whose seconds and nanoseconds we set in unix time 
+ */
+void SetStartTimestamp(const sdk::trace::SpanData* from_span, google::protobuf::Timestamp* proto)
+{
+    std::chrono::nanoseconds unix_time_nanoseconds(from_span->GetStartTime().time_since_epoch().count());
+    auto time_in_seconds = std::chrono::duration_cast<std::chrono::seconds>(unix_time_nanoseconds);
+    proto->set_seconds(time_in_seconds.count());
+    auto time_in_nanoseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(time_in_seconds);
+    proto->set_nanos(unix_time_nanoseconds.count()-time_in_nanoseconds.count());
+}
+
+
+/**
+ * Sets google protobuf object's unix end time, both seconds and nanos
+ * 
+ * @param from_span - The span whose end time we extract and convert to google protobuf timestamp
+ * @param proto - A google protobuf timestamp whose seconds and nanoseconds we set in unix time
+ */
+void SetEndTimestamp(const sdk::trace::SpanData* from_span, google::protobuf::Timestamp* proto)
+{
+    std::chrono::nanoseconds unix_time_nanoseconds(from_span->GetStartTime().time_since_epoch().count()
+                                                    + from_span->GetDuration().count());
+    auto time_in_seconds = std::chrono::duration_cast<std::chrono::seconds>(unix_time_nanoseconds);
+    proto->set_seconds(time_in_seconds.count());
+    auto time_in_nanoseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(time_in_seconds);
+    proto->set_nanos(unix_time_nanoseconds.count()-time_in_nanoseconds.count());
+}
+
+
+/**
+ * Converts each Recordable(Span) to a google protobuf span used by Cloud Trace v2 API
+ * to display traces on GCP
+ * 
+ * @param spans - The list of spans to be exported to google cloud
+ * @param project_id - User specified google cloud project id
+ * @param trace_id_ - *This is subject to change in the future*
+ * @param request - The request we send via gRPC to the cloud to export the traces
+ */
+void ConvertSpans(const nostd::span<std::unique_ptr<sdk::trace::Recordable>> &spans,
+                  const std::string& project_id,
+                  const trace::TraceId& trace_id_,
+                  google::devtools::cloudtrace::v2::BatchWriteSpansRequest* request)
+{
+
+    for (auto &recordable : spans)
+    {
+        auto to_span = request->add_spans();
```

Comment:
The way laid out here is valid, but it's not the most performant. I'd recommend implementing your own `Recordable` derived type that wraps a `to_span`. Any data passed in via `Recordable` methods is then forwarded to `to_span`.

Reply:
The main challenge here would be to make this play nicely with the `BatchWriteSpansRequest` object.

Reply:
I believe I'd need more functionality than is provided by the current [Recordable](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/include/opentelemetry/sdk/trace/recordable.h) because there are a few fields missing that are required for a google protobuf span to be exported.

Reply:
What are you missing in the current recordable? I think SpanData doesn't provide more data than goes through the recordable.

Reply:
Yes, the fields are present, but not custom made for google protobuf objects. And I believe that is expected.

I will have to put a few fields together (for the google protobuf) from the fields of a recordable, and that intermediate work was what I was trying to avoid.

I think I will open a separate PR for recordables to avoid clutter in this PR.